### DISCUSSION TITLE
Upsert with Primary Key error: What am I doing wrong here?

### DISCUSSION THREAD

**Discussion by AdamBridges** [simolus3/drift#1707] [Q&A] [answered]
Hi everyone,

So here's what I'm trying to do:
```
Future insertEntry(ConnectionsTableData entry) =>
      into(db.connectionsTable).insertOnConflictUpdate(entry);
```

And here's what is happening when called from a background task:
```
I/flutter (27101): Drift: Sent INSERT INTO connections_table (uid, name, relationship, date, feeling) VALUES (?, ?, ?, ?, ?) ON CONFLICT(uid) DO UPDATE SET uid = ?, name = ?, relationship = ?, date = ?, feeling = ? with args [emoeOq1fRxOkbd55KJISURHBHi53, Tester, Other, 1645574400000, Bad, emoeOq1fRxOkbd55KJISURHBHi53, Tester, Other, 1645574400000, Bad]
E/BackgroundWorker(27101): errorCode: error, errorMessage: SqliteException(1): ON CONFLICT clause does not match any PRIMARY KEY or UNIQUE constraint, SQL logic error (code 1)
E/BackgroundWorker(27101):   Causing statement: INSERT INTO connections_table (uid, name, relationship, date, feeling) VALUES (?, ?, ?, ?, ?) ON CONFLICT(uid) DO UPDATE SET uid = ?, name = ?, relationship = ?, date = ?, feeling = ?
```
The only thing I can deduce is that the `args` list in the error includes a duplication of my data for some reason but maybe that's fine for how drift operates.

**Any insight here would be appreciated. Thanks!**

Note the following:
```
class ConnectionsTable extends Table {
  TextColumn get uid => text()();
  TextColumn get name => text().nullable()();
  TextColumn get relationship => text().nullable()();
  BlobColumn get photo => blob().nullable()();
  IntColumn get date => integer().map(DateTimeToIntConverter())();
  TextColumn get feeling => text()();

  @override
  Set<Column> get primaryKey => {uid};
}
```
```
dependencies:
  drift: ^1.4.0
  sqlcipher_flutter_libs: ^0.5.1
  path_provider: ^2.0.5
  path: ^1.8.0

dev_dependencies:
  drift_dev: ^1.4.0
  build_runner: ^2.1.7
```
Attempted on emulated Android API 31 and 30 devices and using the Companion class to insert instead with no difference.

**Comment by simolus3** [comment, marked as answer]
The duplicate variables are because we use variables twice (one for `VALUES (?, ?, ?, ?, ?)`, one for `SET uid = ?, name = ?, relationship = ?, date = ?, feeling = ?`). This looks fine to me.

> ON CONFLICT clause does not match any PRIMARY KEY or UNIQUE constraint, SQL logic error

This looks concerning because we do `ON CONFLICT(uid)` and `uid` _should_ be the primary key of the table. Can you enable `logStatements: true` on the `NativeDatabase` constructor and post the `CREATE TABLE` statement for `connections_table` (you'll have to delete your app's data first)? Maybe drift isn't creating the primary key correctly.

**Reply by AdamBridges** [reply]
Okay, I'll try that now.

**Reply by AdamBridges** [reply]
Okay, once the database was recreated, it works fine now. The only items that I can conclude to then that brought about this error are the following:

- column "uid" was originally called "id"
- I later then included a primary key in the table for column "id" and regenerated the *.g.dart file
- Then I renamed column "id" to "uid" and regenerated the *.g.dart file
- Then I migrated the schema to handle the renaming

I'm going to guess that including a Primary Key after a table has been created is the problem? If so, how should that be handled for future reference?

**Reply by simolus3** [reply]
If it's confidential I can also try to reproduce this separately, but do you have a sqlite3 database file after the migration that shows the broken behavior? I'd love to take a look at the `sqlite_schema` table afterwards to check if anything weird is happening. After renaming the column, the conflict clause should be able to refer to the new name.

**Reply by AdamBridges** [reply]
I do and can but it's encrypted by sqlcipher. Is that an issue?

**Reply by simolus3** [reply]
Not if you give me the key too :D Feel free to attach it here or mail it to me at oss@simonbinder.eu.

**Reply by simolus3** [reply]
Ok, so in the schema the primary key is indeed missing, this is what `.schema` gave me:

```sql
CREATE TABLE connections_table (uid TEXT NOT NULL, name TEXT NULL, relationship TEXT NULL, photo BLOB NULL, date INTEGER NOT NULL, feeling TEXT NOT NULL);
```

Given that the problem went away after you re-created the database, I assume that in a previous version of your database code you didn't have the `primaryKey` override? Or perhaps you had it but didn't run `build_runner build` before opening your app? In that case there's a version of the database without the primary key.
If you don't have any released versions of your app where `connections_table` doesn't have a PK, that should be fine. Otherwise, you can bump your `schemaVersion` and write a migration that's essentially `await m.alterTable(TableMigration(connectionsTable))` to re-create the table without data loss.

**Reply by AdamBridges** [reply]
> I assume that in a previous version of your database code you didn't have the primaryKey override?

Correct.

Okay, so just to clarify: if you add a Primary Key after a Table has been created, then you need to recreate the table via a migration, correct?

**Reply by simolus3** [reply]
Yes! In general, _all_ modifications to your tables need an explicit migration to be applied.
